crypto/src/key/hdkd/child_number.rs

```diff
@@ -29,35 +30,78 @@ pub struct ChildNumber(DerivationType);
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
 enum DerivationType {
+    Normal(U31),
     Hardened(U31),
 }
 
 impl ChildNumber {
+    pub const ZERO: Self = ChildNumber(DerivationType::Normal(U31::from_u32_with_msb(0).0));
+
     /// Return a hardened child number
-    pub fn from_hardened(index: U31) -> Self {
+    pub const fn from_hardened(index: U31) -> Self {
         ChildNumber(DerivationType::Hardened(index))
     }
 
+    /// Return a normal child number
+    pub const fn from_normal(index: U31) -> Self {
+        ChildNumber(DerivationType::Normal(index))
+    }
+
     /// Return a child based on the hardened bit (MSB bit)
-    pub fn from_index_with_hardened_bit(index: u32) -> Result<Self, DerivationError> {
+    pub fn from_index_with_hardened_bit(index: u32) -> Self {
         let (index_u31, msb) = U31::from_u32_with_msb(index);
         if msb {
-            Ok(Self::from_hardened(index_u31))
+            Self::from_hardened(index_u31)
         } else {
-            Err(DerivationError::UnsupportedDerivationType)
+            Self::from_normal(index_u31)
         }
     }
 
     /// Return a BIP32-like child number index that has the hardened bit set if needed
     pub fn into_encoded_index(self) -> u32 {
         match self.0 {
+            DerivationType::Normal(i) => i.into_encoded_with_msb(false),
             DerivationType::Hardened(i) => i.into_encoded_with_msb(true),
         }
     }
 
     pub fn into_encoded_be_bytes(self) -> [u8; 4] {
         self.into_encoded_index().to_be_bytes()
     }
+
+    pub fn is_hardened(&self) -> bool {
+        match self.0 {
+            DerivationType::Normal(_) => false,
+            DerivationType::Hardened(_) => true,
+        }
+    }
+
+    pub fn is_normal(&self) -> bool {
+        !self.is_hardened()
+    }
+
+    pub fn plus_one(&self) -> Result<Self, DerivationError> {
+        match self.0 {
+            DerivationType::Normal(i) => Ok(ChildNumber(DerivationType::Normal(i.plus_one()?))),
+            DerivationType::Hardened(i) => Ok(ChildNumber(DerivationType::Hardened(i.plus_one()?))),
+        }
+    }
+}
+
+impl Encode for ChildNumber {
```

Comment:
Please write tests for Encoding/Decoding that are specific to ChildNumber in this module. Also, with ChildNumber becoming more complex, we should test all its functionality. Please add unit tests for the life cycle of ChildNumber.

Reply:
To be clear, always in encoding/decoding tests, we test:

1. Back and forth encoding with random values
2. Decoding from selected values, to ensure that the underlying format is stable

Reply:
I'm sorry, I still don't see these tests.